File: backend/app/api/routes/adapter.py
"""
Typing adaptation engine endpoints.
Exposes adapter recommendations and configuration.
"""

from fastapi import APIRouter, Depends

from app.api.schemas import AdapterConfig, AdapterRecommendation
from app.db.database import get_db
from sqlalchemy.orm import Session

router = APIRouter(prefix="/adapter")


@router.get("/config", response_model=AdapterConfig)
async def get_adapter_config(db: Session = Depends(get_db)):
    """
    Get current adapter configuration.
    
    Returns the active adapter type and its parameters.
    """
    # TODO: Get user from token
    # TODO: Get user's adapter preference
    # TODO: Return config
    pass


@router.put("/config", response_model=AdapterConfig)
async def update_adapter_config(
    config: AdapterConfig,
    db: Session = Depends(get_db),
):
    """
    Update user's adapter configuration.
    
    Allows switching between adapters or adjusting parameters.
    """
    # TODO: Get user from token
    # TODO: Validate adapter type
    # TODO: Store user preference
    # TODO: Return updated config
    pass


@router.post("/recommend", response_model=AdapterRecommendation)
async def get_recommendation(
    session_id: int,
    db: Session = Depends(get_db),
):
    """
    Get adapter recommendation for next session.
    
    Based on:
    - User's recent performance
    - Current difficulty level
    - Error patterns
    - Improvement trends
    
    Supports multiple adapters:
    - Rule-based: Simple heuristics
    - ML: Machine learning predictions (future)
    """
    # TODO: Get user from token
    # TODO: Load recent session data
    # TODO: Invoke appropriate adapter
    # TODO: Return recommendation with confidence
    pass


@router.get("/available-adapters")
async def get_available_adapters():
    """
    Get list of available adapters and their descriptions.
    """
    # TODO: Return list of registered adapters
    # - rule_based: Description
    # - ml: Description
    # - custom: Any custom adapters
    pass

File: frontend/src/services/adapter.ts
/**
 * Adapter communication service
 * Handles interaction with the typing adaptation engine
 */

import { apiClient } from './api'
import { AdapterRecommendation } from '@types/index'

export const adapterService = {
  /**
   * Get current adapter configuration
   */
  async getConfig() {
    return apiClient.getAdapterConfig()
  },

  /**
   * Update adapter configuration
   */
  async updateConfig(config: Record<string, unknown>) {
    return apiClient.updateAdapterConfig(config)
  },

  /**
   * Get difficulty recommendation for next session
   */
  async getRecommendation(sessionId: number): Promise<AdapterRecommendation> {
    return apiClient.getRecommendation(sessionId)
  },

  /**
   * Calculate difficulty based on performance
   */
  calculateDifficulty(wpm: number, accuracy: number): number {
    if (accuracy < 70) return 1
    if (wpm < 30) return 1
    if (wpm < 40) return 2
    if (wpm < 50) return 3
    if (wpm < 60) return 4
    if (wpm < 70) return 5
    if (wpm < 80) return 6
    if (wpm < 90) return 7
    return 8
  },

  /**
   * Get focus areas based on performance
   */
  getFocusAreas(accuracy: number, wpm: number): string[] {
    const areas = []
    if (accuracy < 85) areas.push('Accuracy')
    if (wpm < 50) areas.push('Speed')
    if (areas.length === 0) areas.push('Push for higher difficulty')
    return areas
  },
}
